---
title: "GSEA figures"
author: "Mikhail Dozmorov"
date: "`r Sys.Date()`"
output:
  pdf_document:
    toc: no
  html_document:
    theme: cerulean
    toc: yes
---

```{r setup, echo=FALSE, message=FALSE, warning=FALSE}
# Set up the environment
library(knitr)
opts_chunk$set(cache.path='cache/', fig.path='img/', cache=F, tidy=T, fig.keep='high', echo=F, dpi=100, warnings=F, message=F, comment=NA, warning=F, results='as.is', fig.width = 10, fig.height = 6) #out.width=700, 
library(pander)
panderOptions('table.split.table', Inf)
set.seed(1)
library(dplyr)
options(stringsAsFactors = FALSE)
```

```{r libraries}
# Libraries

library(readxl)
library(cowplot)
library(ggplot2)
library(stringr)
library(writexl)
library("ggsci")
library(scales)
# scales::show_col(pal_lancet("lanonc")(8))
mycols = pal_lancet("lanonc")(8)
```

```{r settings}
# Settings

# Data
data_dir       <- "/Users/mdozmorov/Documents/Work/VCU_work/Ginder/2023-05-16.GATAD2_CHD4" # Working directory
p_adj_cutoff   <- 0.05 # FDR cutoff
fileNameIn1 <- file.path(data_dir, paste0("results/GSEA_DEGs_", p_adj_cutoff, "_GATAD2B.xlsx"))
fileNameIn2 <- file.path(data_dir, paste0("results/GSEA_DEGs_", p_adj_cutoff, "_CHD4.xlsx"))
# Save integrated results
fileNameOut1 <- file.path(data_dir, "results/GSEA_combined_GATAD2B_CHD4.xlsx")
# Create output folder, if not exists
if (!dir.exists(file.path(data_dir, "results"))) dir.create(file.path(data_dir, "results"))
msigdb_all     <- FALSE # Use all MSigDb categories (TRUE), or c("C2", "C5", "H") (FALSE)

max_GO_to_plot = 20   # Maximum number of GOs to plot
max_enrichment_length <- 50 # Maximum length of enrichment descriptions
```

# Visualization

- Two types of Enrichment analyses were run:
    - **Enrich** - enrichment _without considering_ fold change of DEGs
    - **GSEA** - enrichment _considering_ the order of up- and downregulated DEGs. **NES** - normalized enrichment score - if positive, the process is enriched in UPregulate genes, and vice versa.

- The following signatures were used:
    - **KEGG** - KEGG pathways
    - MSigDB collections, see  https://www.gsea-msigdb.org/gsea/msigdb/index.jsp. 
        - **H**	hallmark gene sets  are coherently expressed signatures derived by aggregating many MSigDB gene sets to represent well-defined biological states or processes.
        - **C1**	positional gene sets  for each human chromosome and cytogenetic band.
        - **C2**	curated gene sets  from online pathway databases, publications in PubMed, and knowledge of domain experts.
        - **C3**	regulatory target gene sets  based on gene target predictions for microRNA seed sequences and predicted transcription factor binding sites.
        - **C4**	computational gene sets  defined by mining large collections of cancer-oriented microarray data.
        - **C5**	GO gene sets  consist of genes annotated by the same GO terms.
        - **C6**	oncogenic gene sets  defined directly from microarray gene expression data from cancer gene perturbations.
        - **C7**	immunologic gene sets  defined directly from microarray gene expression data from immunologic studies.
- Sheets in the Excel results file are named by the **type of analysis**.**signature**, e.g., **Enrich.KEGG**, **GSEA.C7**

**Signatures in question can be googled**

## -log10 p-value interpretation

$$1 = 10^{-1} = 0.1$$
$$1.3 = 10^{-1.3} \approx 0.05$$
$$2 = 10^{-2} = 0.01$$
$$4 = 10^{-4} = 0.0001$$
etc. -log10 adjusted p-value $\ge1$ is acceptable

```{r fig.height=4}
# Sheets depending on what analyses were run
if (msigdb_all) {
  all_sheets <- excel_sheets(fileNameIn1)
} else {
  all_sheets <- c("Enrich.KEGG", "GSEA.KEGG", "Enrich.C2", "GSEA.C2", "Enrich.C5", "GSEA.C5", "Enrich.H", "GSEA.H")
}

mtx_summaries <- list()

# Go through each sheet
for (sheet in all_sheets) {
  # Analysis type
  print(sheet)
  # Process Enrichr results
  if (sheet == "Enrich.KEGG") {
    mtx1 <- read_xlsx(fileNameIn1, sheet = sheet)
    mtx2 <- read_xlsx(fileNameIn2, sheet = sheet)
    # Add -log10 adjusted p-value
    mtx1$log10pval <- -log10(mtx1$P.value)
    mtx2$log10pval <- -log10(mtx2$P.value)
    mtx <- full_join(mtx1, mtx2, by = c("Term"), suffix = c(".GATAD2B", ".CHD4")) %>% dplyr::select(Term, log10pval.GATAD2B, log10pval.CHD4) %>% dplyr::arrange(desc(log10pval.GATAD2B))
    # # Add ranks
    # mtx$ranklog10pval.GATAD2B <- rank(mtx$log10pval.GATAD2B)
    # mtx$ranklog10pval.CHD4 <- rank(mtx$log10pval.CHD4)
    # mtx$rankaverage <- rowMeans(mtx[, c("ranklog10pval.GATAD2B", "ranklog10pval.CHD4")], na.rm = TRUE)
    # mtx <- mtx[order(mtx$rankaverage, decreasing = TRUE), ]
    mtx <- mtx[order(mtx$log10pval.GATAD2B, decreasing = TRUE), ]
    # Save summary
    mtx_summaries <- c(mtx_summaries, list(mtx))
    names(mtx_summaries)[length(mtx_summaries)] <- sheet
    if (nrow(mtx) > 0) {
      mtx <- mtx[1:min(nrow(mtx), max_GO_to_plot), ]
    }
    # Add what name to plot
    mtx$enrichment <- str_trunc(mtx$Term, width = max_enrichment_length)
    # Make long format
    mtx <- tidyr::pivot_longer(mtx[, c("enrichment", "log10pval.GATAD2B", "log10pval.CHD4")], cols = c(log10pval.GATAD2B, log10pval.CHD4))
  } else if (sheet == "GSEA.KEGG") {
    # Process GSEA results
    mtx1 <- read_xlsx(fileNameIn1, sheet = sheet)
    mtx2 <- read_xlsx(fileNameIn2, sheet = sheet)
    # Add -log10 adjusted p-value
    mtx1$log10pval <- -log10(as.numeric(mtx1$pvalue))
    mtx2$log10pval <- -log10(as.numeric(mtx2$pvalue))
    mtx <- full_join(mtx1, mtx2, by = c("Description"), suffix = c(".GATAD2B", ".CHD4")) %>% dplyr::select(Description, log10pval.GATAD2B, log10pval.CHD4) %>% dplyr::arrange(desc(log10pval.GATAD2B))
    # # Add ranks
    # mtx$ranklog10pval.GATAD2B <- rank(mtx$log10pval.GATAD2B)
    # mtx$ranklog10pval.CHD4 <- rank(mtx$log10pval.CHD4)
    # mtx$rankaverage <- rowMeans(mtx[, c("ranklog10pval.GATAD2B", "ranklog10pval.CHD4")], na.rm = TRUE)
    # mtx <- mtx[order(mtx$rankaverage, decreasing = TRUE), ]
    mtx <- mtx[order(mtx$log10pval.GATAD2B, decreasing = TRUE), ]
    # Save summary
    mtx_summaries <- c(mtx_summaries, list(mtx))
    names(mtx_summaries)[length(mtx_summaries)] <- sheet
    if (nrow(mtx) > 0) {
      mtx <- mtx[1:min(nrow(mtx), max_GO_to_plot), ]
    }
    # Add what name to plot
    mtx$enrichment <- str_trunc(mtx$Description, width = max_enrichment_length)
    # Make long format
    mtx <- tidyr::pivot_longer(mtx[, c("enrichment", "log10pval.GATAD2B", "log10pval.CHD4")], cols = c(log10pval.GATAD2B, log10pval.CHD4))
  } else {
    # Process MSigDb data
    mtx1 <- read_xlsx(fileNameIn1, sheet = sheet)
    mtx2 <- read_xlsx(fileNameIn2, sheet = sheet)
    # Add -log10 adjusted p-value
    mtx1$log10pval <- -log10(as.numeric(mtx1$pvalue))
    mtx2$log10pval <- -log10(as.numeric(mtx2$pvalue))
    # Change the direction, if negative enrichment score
    if (grepl("GSEA", sheet)) {
       mtx1$log10pval <- mtx1$log10pval * ifelse(mtx1$NES > 0, 1, -1)
       mtx2$log10pval <- mtx2$log10pval * ifelse(mtx2$NES > 0, 1, -1)
    }
    mtx <- full_join(mtx1, mtx2, by = c("Description"), suffix = c(".GATAD2B", ".CHD4")) %>% dplyr::select(Description, log10pval.GATAD2B, log10pval.CHD4) %>% dplyr::arrange(desc(log10pval.GATAD2B))
    # # Add ranks
    # mtx$ranklog10pval.GATAD2B <- rank(mtx$log10pval.GATAD2B)
    # mtx$ranklog10pval.CHD4 <- rank(mtx$log10pval.CHD4)
    # mtx$rankaverage <- rowMeans(mtx[, c("ranklog10pval.GATAD2B", "ranklog10pval.CHD4")], na.rm = TRUE)
    # mtx <- mtx[order(mtx$rankaverage, decreasing = TRUE), ]
    mtx <- mtx[order(abs(mtx$log10pval.GATAD2B), decreasing = TRUE), ]
    # Save summary
    mtx_summaries <- c(mtx_summaries, list(mtx))
    names(mtx_summaries)[length(mtx_summaries)] <- sheet
    if (nrow(mtx) > 0) {
      mtx <- mtx[1:min(nrow(mtx), max_GO_to_plot), ]
    }
    # Add what name to plot
    mtx$enrichment <- str_trunc(mtx$Description, width = max_enrichment_length)
    # Make long format
    mtx <- tidyr::pivot_longer(mtx[, c("enrichment", "log10pval.GATAD2B", "log10pval.CHD4")], cols = c(log10pval.GATAD2B, log10pval.CHD4))
  }
  
  # Actual plotting
  
  p <- ggplot(data = mtx, aes(x = reorder(enrichment, value), y = value)) +
    geom_col(aes(fill = name), position=position_dodge()) +
#    scale_fill_gradient2(low = "blue", high = "red", midpoint = 0) +
    scale_fill_manual(values = rev(mycols[2:1]), name = "") + # get rid of legend title
    coord_flip() +
    theme_cowplot() +
    ylab("-log10 p-value") +
    xlab("Enrichment term")
  print(p)
}
```

# Save the data

Results are saved in `r basename(fileNameOut1)`. Columns "log10pval.GATAD2B" and "log10pval.CHD4" show enrichment of a pathway in the corresponding condition. Large _absolute_ values indicate stronger enrichment. Negative sign indicates a pathway is downregulated. NA indicate the pathway is not enriched.

```{r}
x <- mtx_summaries
write_xlsx(x, fileNameOut1)
```

